tests/net/socket/tls/src/main.c

```diff
@@ -0,0 +1,244 @@
+#include <posix/unistd.h>
+#include <ztest.h>
+
+LOG_MODULE_REGISTER(tls_test, CONFIG_NET_SOCKETS_LOG_LEVEL);
+
+#define STACK_SIZE 1024
+
+#define PORT 4242
+
+/* The answer to life, the universe, and everything */
+#define SECRET "forty-two"
+#define SECRET_SIZE 9
+
+enum tls_tag {
+	SERVER_CERTIFICATE_TAG,
+};
+
+/** @brief The server thread stack */
+static K_THREAD_STACK_DEFINE(server_stack, STACK_SIZE);
+/** @brief the server thread object */
+static struct k_thread server_thread;
+
+/**
+ * @brief The Certificate Authority (CA) Certificate
+ *
+ * This is usually only required if the server is using
+ * a self-signed certificate.
+ */
+static const unsigned char ca[] = {
+#include "ca.inc"
+};
+
+/**
+ * @brief The Server Certificate
+ *
+ * This is the public key of the server.
+ */
+static const unsigned char server[] = {
+#include "server.inc"
+};
+
+/**
+ * @brief The Server Private Key
+ *
+ * This is the public key of the server.
+ */
+static const unsigned char server_privkey[] = {
+#include "server_privkey.inc"
+};
+
+/**
+ * @brief The server thread function
+ *
+ * This function simply accepts a client connection and
+ * echoes the first @ref SECRET_SIZE bytes of the first
+ * packet. After that, the server is closed and connections
+ * are no longer accepted.
+ *
+ * @param arg0 a pointer to the int representing the server filed descriptor
+ * @param arg1 ignored
+ * @param arg2 ignored
+ */
+static void server_thread_fn(void *arg0, void *arg1, void *arg2)
+{
+	const int server_fd = (int)(intptr_t)arg0;
+	int r;
+	int client_fd;
+	socklen_t addrlen;
+	char addrstr[INET_ADDRSTRLEN];
+	struct sockaddr_in sa;
+	char *addrstrp;
+
+	NET_DBG("Server thread running");
+
+	memset(&sa, 0, sizeof(sa));
+    addrlen = sizeof(sa);
+
+    NET_DBG("Accepting client connection..");
+    r = accept(server_fd, (struct sockaddr *)&sa, &addrlen);
+    __ASSERT(r != -1, "accept() failed (%d)", r);
+    client_fd = r;
+
+    memset(addrstr, '\0', sizeof(addrstr));
+    addrstrp = (char *)inet_ntop(AF_INET, &sa.sin_addr,
+        addrstr, sizeof(addrstr));
+    __ASSERT(addrstrp != NULL, "inet_ntop() failed (%d)", errno);
+
+    NET_DBG("accepted connection from [%s]:%d as fd %d",
+        log_strdup(addrstr), ntohs(sa.sin_port), client_fd);
+
+    NET_DBG("calling recv()");
+    r = recv(client_fd, addrstr, sizeof(addrstr), 0);
+    __ASSERT(r != -1, "recv() failed (%d)", errno);
+    __ASSERT(r == SECRET_SIZE, "expected: %u actual: %u", SECRET_SIZE, r);
+
+    NET_DBG("calling send()");
+    r = send(client_fd, SECRET, SECRET_SIZE, 0);
+	__ASSERT(r != -1, "send() failed (%d)", errno);
+    __ASSERT(r == SECRET_SIZE, "expected: %u actual: %u", SECRET_SIZE, r);
+
+    NET_DBG("closing client fd");
+    r = close(client_fd);
+    __ASSERT(r != -1, "close() failed on the server fd (%d)", errno);
+}
+
+static void test_tls(void)
+{
+	const int yes = true;
+	static const sec_tag_t sec_tag_list[] = {
+		SERVER_CERTIFICATE_TAG,
```

Comment:
It isn't a good idea to put all credentials together, `ca` should only be used on the client side, `server` and `server_privkey` on the server side. I'd create a sperate tag for the CA certificate and use it with the client instead. Server doesn't need to register CA either as it does not authenticate the client.

If you want to test mutual authentication it would rather make sense to create a separate set of private key and certificate for the client and the server.

Reply:
@rlubos That's a good point - ultimately, I would like to provide an example for client authentication as well.

I ended up starting the server automatically with `K_THREAD_DEFINE()` and moved the server initialisation there.

It might be good to have ~3 test variants:
* 1 where TLS is not used (sure yea, this may be unnecessary, but it's a good base case)
* 1 where only the client authenticates the server (server needs server certs only)
* 1 with mutual authentication (server needs server certs + ca cert)

Reply:
That reminds me of my next question - how does the client specify its public key for when the server has to verify the client with `TLS_PEER_VERIFY_REQUIRED`.

Reply:
> 1 where TLS is not used (sure yea, this may be unnecessary, but it's a good base case)

Not sure if you need to bother about this one - regular TCP sockets are already covered bu unit test.

> That reminds me of my next question - how does the client specify its public key for when the server has to verify the client with `TLS_PEER_VERIFY_REQUIRED`

It should work the same way as on the server side - by registering `TLS_CREDENTIAL_PRIVATE_KEY`/`TLS_CREDENTIAL_SERVER_CERTIFICATE` (yeah, I know, the naming isn't very fortunate in this case, would be better to have it called `TLS_CREDENTIAL_OWN_CERTIFICATE` actually...) and setting them via `setsockopt` on the socket (by using corresponding tag). The server should register a corresponding CA certificate (it's the same cert in case it's self-signed).
Inernally we'll end up calling `mbedtls_ssl_conf_own_cert()` which can register certificate/key for both client and server.

Reply:
> > That reminds me of my next question - how does the client specify its public key for when the server has to verify the client with `TLS_PEER_VERIFY_REQUIRED`
> 
> It should work the same way as on the server side - by registering `TLS_CREDENTIAL_PRIVATE_KEY`/`TLS_CREDENTIAL_SERVER_CERTIFICATE` (yeah, I know, the naming isn't very fortunate in this case, would be better to have it called `TLS_CREDENTIAL_OWN_CERTIFICATE` actually...) and setting them via `setsockopt` on the socket (by using corresponding tag). The server should register a corresponding CA certificate (it's the same cert in case it's self-signed).
> Inernally we'll end up calling `mbedtls_ssl_conf_own_cert()` which can register certificate/key for both client and server.

I see ok. I always like to relate things to SSH, so I'm wondering, if there were something analogous to `~/.ssh/authorized_keys` how would that work with TLS certs? Is it equivalent to checking an arbitrary certificate against the CA public key for peer verification?

Reply:
If client certifiacates are all self-signed, then you can register all those certs in your server to allow the clients to connect. But it's cumbersome. IMO it would be better to have one trusted certificate registered on the server, and use that cert (or rather it's private key) to sign your client certificates. The effect would would be the same - only the devices that have certificates signed by your trusted one will be able to connect to your server.

But I believe that any resource in the internet could explain this better than me - just look for "Public Key Infrastrucutre (PKI)"